fs-repo-14-to-15/migration/migration.go

```diff
@@ -0,0 +1,248 @@
+		log.Error("failed to update version file to 15")
+		// There was an error so abort writing the output and clean up temp file
+		panicOnError(out.Abort())
+		panicOnError(backup.Abort())
+		return err
+	} else {
+		// Write the output and clean up temp file
+		panicOnError(out.Close())
+		panicOnError(backup.Close())
+	}
+
+	log.Log("updated version file")
+
+	log.Log("Migration 14 to 15 succeeded")
+	return nil
+}
+
+// panicOnError is reserved for checks we can't solve transactionally if an error occurs
+func panicOnError(e error) {
+	if e != nil {
+		panic(fmt.Errorf("error can't be dealt with transactionally: %w", e))
+	}
+}
+
+func (m Migration) Revert(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("reverting migration")
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+	if err := repo.CheckVersion("15"); err != nil {
+		return err
+	}
+
+	cfg := filepath.Join(opts.Path, "config")
+	if err := os.Rename(cfg+backupSuffix, cfg); err != nil {
+		return err
+	}
+
+	if err := repo.WriteVersion("14"); err != nil {
+		return err
+	}
+	if opts.Verbose {
+		log.Log("lowered version number to 14")
+	}
+
+	return nil
+}
+
+// convert converts the config from one version to another
+func convert(in io.Reader, out io.Writer) error {
+	confMap := make(map[string]any)
+	if err := json.NewDecoder(in).Decode(&confMap); err != nil {
+		return err
+	}
+
+	// Upgrade bootstrapper QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ from /quic to /quic-v1
+	if b, ok := confMap["Bootstrap"]; ok {
+		bootstrap, ok := b.([]interface{})
+		if !ok {
+			return fmt.Errorf("invalid type for .Bootstrap got %T expected json array", b)
+		}
+
+		for i, v := range bootstrap {
+			if v == "/ip4/104.131.131.82/udp/4001/quic/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ" {
+				bootstrap[i] = "/ip4/104.131.131.82/udp/4001/quic-v1/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"
+			}
+		}
+	}
+
+	// Remove /quic only addresses from .Addresses.Swarm
+	if a, ok := confMap["Addresses"]; ok {
+		addresses, ok := a.(map[string]any)
+		if !ok {
+			return fmt.Errorf("invalid type for .Addresses got %T expected json map", a)
+		}
+
+		if s, ok := addresses["Swarm"]; ok {
+			swarm, ok := s.([]interface{})
+			if !ok {
+				return fmt.Errorf("invalid type for .Addresses.Swarm got %T expected json array", s)
+			}
+
+			var newSwarm []interface{}
+			for _, v := range swarm {
+				if addr, ok := v.(string); ok {
+					if !strings.Contains(addr, "/quic") || strings.Contains(addr, "/quic-v1") {
+						newSwarm = append(newSwarm, addr)
+					}
+				} else {
+					newSwarm = append(newSwarm, v)
+				}
+			}
+			addresses["Swarm"] = newSwarm
+		}
+	}
+
+	// Remove legacy Gateway.HTTPHeaders values that were hardcoded since years ago, but no longer necessary
+	// (but leave as-is if user made any changes)
+	// https://github.com/ipfs/kubo/issues/10005
```

Comment:
:information_source: Added cleanup of `Gateway.HTTPHeaders`.  It was a long time due. 
Details and rationale in  https://github.com/ipfs/kubo/issues/10005